### DISCUSSION TITLE
Extending the Arithmetic example

### DISCUSSION THREAD

**Discussion by alirazeen** [congo-cc/congo-parser-generator#184] [Q&A]
The arithmetic example is a great way to introduce new users to congo-cc. There are lots of new concepts to learn (defining the grammar, learning the grammar syntax, learning how to actually use the generated parsers, etc.). I imagine that a tutorial that builds these things up incrementally will be useful. For example, something that shows the following step-by-step:

- Adding support for arithmetic expressions such as `-1 + 2`.
- Adding support for mathematical operators like `!` and `^`.
- Using the parsers in their own program instead of executing the `Main()` function directly.

However, I have a question that I don't know how to answer. How will I add support for the expression `+1 + 2`? Adding support for `-1 + 2` is easy, I just have to do this:

```
<NUMBER :  ("-")?(["0"-"9"])+ ("."(["0"-"9"])+)?>
```

But I can't do something similar to support `+1 + 2`. I can't do this, for example:

```
<NUMBER :  ("-"|"+")?(["0"-"9"])+ ("."(["0"-"9"])+)?>
```

If I do this, then the "+ 2" in `-1 + 2` will be parsed as `NUMBER` when really, the second `+` should be an operator.

**Comment by revusky** [comment]
I don't think you can add support for `+` and `-` as unary operators on the purely lexical level. You would need to do it at the syntactic level. I guess the most informative thing would be to consult the various more complex grammars. For example, see: https://github.com/congo-cc/congo-parser-generator/blob/a1c69f369463ba4b4fb9b3ef792b01a6d72b06b0/examples/java/Java.ccc#L767 or: https://github.com/congo-cc/congo-parser-generator/blob/a1c69f369463ba4b4fb9b3ef792b01a6d72b06b0/examples/python/Python.ccc#L237

Regardless, one could definitely extend the tutorial, having the toy grammar fleshed out with more language features. Though, that said, it's not so obvious at what point, you just say to the reader: "Well, if you want to see this kind of thing extended in a non-toy example, here are some grammars for lua/python/java etcetera that you can consult."

Of course, the more meaty examples with full programming languages are maybe less pedagogical, since, with a more contrived toy example, you can keep this cleaner and simpler. These real-world langauges that evolved over time typically have a lot of messy corners in the syntax.

**Comment by alirazeen** [comment]
Yes, that should work. I kept thinking there was some other trick I could be using but this is cleaner. Okay, this makes sense. Thank you!
-------- Original Message --------
